### PR TITLE
Bimorph mirrors

### PR DESCRIPTION
**Pull request by dan-fernandes**
Adds the `BimorphMirror` device, representing an n-channel CAENels Bimorph Mirror.

### My general questions:
1. Some tests have assertions inside `for` loops. Could these be improved by collecting results and asserting all of collection?
2. Should the mirror throw an error on being instantiated with zero (or negative) channels, or could there be a use case for this? (Perhaps not a programming question.)
3. The test script contains a `valid_bimorph_values` fixture that generates a prandom set of correct inputs. Does it matter that this is generated at collection? Could there be a test case where multiple sets of randomly generated inputs are require, and in this case would it be better to implement input generation using a factory pattern?
4. Order of fixture execution is based on scope, dependency and autouse. Due to this, the `mirror` fixture will definitely be executed before its needed. However, only one reference to the `mirror` fixtures passes the required `number_of_channels` parameter. Is this an issue? (Does this question make sense?)

### Instructions to reviewer on how to test:
1. Ensure tests pass

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://diamondlightsource.github.io/dodal/main/reference/device-standards.html)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly
- [ ] Have the connection tests for the relevant beamline(s) been run via `dodal connect ${BEAMLINE}`

### CODE REVIEW

**Comment by DiamondJoseph** [comment]
Some tests have assertions inside for loops. Could these be improved by collecting results and asserting all of collection?
- The tests that are asserting inside the for loops are iterating over the dictionary that is passed in. 
```python
    for i in range(1, mirror.number_of_channels):
        assert (
            read[f"{BIMORPH_NAME}-channels-{i}-vout"]["value"]
            == valid_bimorph_values[i]
        )
# You could instead, with a dict comprehension:
assert {name: value["value"] for name, value in read.items()} == {f"{bimorph_mirror.name}-channels-{i}-vout": value for i, value in valid_bimorph_values.items()}
```
Should the mirror throw an error on being instantiated with zero (or negative) channels, or could there be a use case for this? (Perhaps not a programming question.)
- I think we should guard against the non-physical negative channels, but 0 is valid. It looks like currently the default argument is 0, but I would make it a required argument (move it before name and don't give it a default).

The test script contains a valid_bimorph_values fixture that generates a prandom set of correct inputs. Does it matter that this is generated at collection? Could there be a test case where multiple sets of randomly generated inputs are require, and in this case would it be better to implement input generation using a factory pattern?
- Grumble grumble I hate fun no random numbers. Right now these numbers are being used in the setup of a test to make sure that the output of the signals will match the input of the set signals. But this means the input is being set prior to the output being set, and your test that asserts that the output is correct would work even if you removed the set method! Adding a side effect on the mock put (as discussed) to propagate from VTARGT->VOUT when the trigger is processed would make the test more representative, and mean you don't need this fixture. I would instead have 2-3 tests that set a subset of the signals, and assert that those signals change and that other signals do not change.

Order of fixture execution is based on scope, dependency and autouse. Due to this, the mirror fixture will definitely be executed before its needed. However, only one reference to the mirror fixtures passes the required number_of_channels parameter. Is this an issue? (Does this question make sense?)
- As discussed, this isn't an issue because Pytest is finding the fixtures and resolving them in an order that ensures that the mirror has been created with the correct number of channels prior to being passed into the other fixtures and the test.

**Comment by dan-fernandes** [comment]
Tests not passing while waiting on this PR: https://github.com/DiamondLightSource/dodal/pull/952

**Comment by dan-fernandes** [comment]
Updates made per Joseph's comment.

File: tests/devices/unit_tests/test_bimorph_mirror.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+from ophyd_async.testing import get_mock_put
+
+from dodal.devices.bimorph_mirror import BimorphMirror
+
+VALID_BIMORPH_CHANNELS = [8, 12, 16, 24]
+
+
+@pytest.fixture
+def mirror(request, RE: RunEngine) -> BimorphMirror:
+    number_of_channels = request.param
```
Comment (callumforrester) [inline tests/devices/unit_tests/test_bimorph_mirror.py:15]: Are you using `request.param` or is the number of channels constant?

File: src/dodal/devices/bimorph_mirror.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        Args:
+            prefix: str PV prefix
+            number_of_channels: int number of channels on bimorph mirror (can be zero)
+            name: str name of device
+
+        Raises:
+            ValueError: number_of_channels is less than zero"""
+
+        if number_of_channels < 0:
+            raise ValueError(f"Number of channels is below zero: {number_of_channels}")
```
Comment (callumforrester) [inline src/dodal/devices/bimorph_mirror.py:81]: Could: Is there a valid use case for 0 channels? Maybe we should limit it to 1?

File: src/dodal/devices/bimorph_mirror.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+        self.number_of_channels = number_of_channels
```
[the remark this answers is not in the document]
Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py:83]: That makes sense. Should the list of channels be an immutable tuple as well?

[the remark this answers is not in the document]
Comment (callumforrester) [reply src/dodal/devices/bimorph_mirror.py]: I think that matters less because the `DeviceVector` constructor will make them immutable anyway

**Review by callumforrester — APPROVED** [review]

File: src/dodal/devices/bimorph_mirror.py (hunk 2 above)
Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py:81]: There are PVs on the mirror that are not in the channel. There is a use case for accessing these without needing access to the channels.

Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py:81]: Added test with zero channels.

File: src/dodal/devices/bimorph_mirror.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            )
+        self.on_off = epics_signal_w(BimorphMirrorOnOff, f"{prefix}ONOFF")
+        self.alltrgt_proc = epics_signal_x(f"{prefix}ALLTRGT.PROC")
+        self.status = epics_signal_r(BimorphMirrorStatus, f"{prefix}STATUS")
+        self.err = epics_signal_r(str, f"{prefix}ERR")
+
+        super().__init__(name=name)
+
+    @AsyncStatus.wrap
+    async def set(self, value: dict[int, float]):
```
[the remark this answers is not in the document]
Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py]: After discussion with Diamond Joseph I have decided to leave this as-is for now.

While there are multiple ways to set a bimorph mirror, this seems the most logical for this type PSU.

**Review by callumforrester — APPROVED** [review]
LGTM, thanks

File: src/dodal/devices/bimorph_mirror.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py]: This logic did not work. I have replaced it.

File: src/dodal/devices/bimorph_mirror.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+        Args:
+            value: Dict of channel numbers to target voltages
+
+        Raises:
+            ValueError: On set to non-existent channel"""
+
+        for i in value.keys():
+            if self.channels.get(i) is None:
+                raise ValueError(f"Set to non-existent channel: {i}")
```
[the remark this answers is not in the document]
Comment (dan-fernandes) [reply src/dodal/devices/bimorph_mirror.py]: This logic did not work. I've replaced it.

**Pull request merged**